src/Cake.Core.Tests/Unit/CakeEngineTests.cs

```diff
@@ -1000,5 +1020,399 @@ public void Should_Return_Report_That_Marks_Delegated_Tasks_As_Delegated()
                 Assert.Equal(CakeTaskExecutionStatus.Delegated, report.First(e => e.TaskName == "B").ExecutionStatus);
             }
         }
+
+        public sealed class TheSetupEvent
+        {
+            [Fact]
+            public void Should_Raise_Setup_Event()
+            {
+                // Given
+                var fixture = new CakeEngineFixture();
+                var engine = fixture.CreateEngine();
+                engine.RegisterTask("A");
+
+                // When
+                var result = Assert.Raises<SetupEventArgs>(
+                    handler => engine.Setup += handler,
+                    handler => engine.Setup -= handler,
+                    () => engine.RunTarget(fixture.Context, fixture.ExecutionStrategy, "A"));
+
+                // Then
+                Assert.NotNull(result);
+                Assert.Equal(engine, result.Sender);
+                Assert.Equal(fixture.Context, result.Arguments.Context);
+            }
+
+            [Fact]
+            public void Should_Invoke_All_Handlers()
+            {
+                // Given
+                var list = new List<ICakeRuntime>();
+                var fixture = new CakeEngineFixture();
+                var engine = fixture.CreateEngine();
+                engine.RegisterTask("A");
+                engine.Setup += (sender, args) =>
+                {
+                    list.Add(args.Context.Environment.Runtime);
+                };
+                engine.Setup += (sender, args) =>
+                {
+                    list.Add(args.Context.Environment.Runtime);
+                };
+
+                // When
+                engine.RunTarget(fixture.Context, fixture.ExecutionStrategy, "A");
+
+                // Then
+                Assert.Equal(2, list.Count);
+                Assert.All(list, runtime => Assert.Equal(fixture.Environment.Runtime, runtime));
```

Comment:
Couldn't we make the two different handlers add two different values and make sure the two of them exist. This test could show false positives if a bug would make the Setup event to be invoked twice instead of once.